File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.cpp
class Solution {
private:
    map<char,vector<char>> hashMap = {{'2', {'a','b','c'}},
                                    {'3', {'d','e','f'}},
                                    {'4', {'g','h','i'}},
                                    {'5', {'j','k','l'}},
                                    {'6', {'m','n','o'}},
                                    {'7', {'p','q','r','s'}},
                                    {'8', {'t','u','v'}},
                                    {'9', {'w','x','y','z'}}};
public:
    void dfs(string digits, string& tmp, vector<string>& answer, int index){
        if(tmp.length() >= digits.length()){
            answer.push_back(tmp);
            return; 
        }

        for(int i = 0; i < hashMap[digits[index]].size(); i++){
            tmp += hashMap[digits[index]][i]; 
            dfs(digits,tmp,answer,index+1); 
            tmp.pop_back();
        }
    }
    vector<string> letterCombinations(string digits) {
        if(digits.empty()) return {}; 
        vector<string> answer; 
        string tmp = ""; 
        dfs(digits,tmp,answer,0); 
        return answer; 
    }
};

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.java
class Solution {
        private final Map<Character,List<Character>> hashMap = Map.of('2', Arrays.asList('a','b','c'), '3', Arrays.asList('d','e','f'), '4', Arrays.asList('g','h','i'), '5', Arrays.asList('j','k','l'), '6', Arrays.asList('m','n','o'), '7', Arrays.asList('p','q','r','s'), '8', Arrays.asList('t','u','v'), '9', Arrays.asList('w','x','y','z')); 
    public List<String> letterCombinations(String digits) {
        List<String> answer = new ArrayList<>(); 
        StringBuilder sb = new StringBuilder(); 
        dfs(digits,answer,0,sb);
        return answer; 
    }
    public void dfs(String digits, List<String> answer, int index, StringBuilder sb){
        if(index >= digits.length()){
            answer.add(sb.toString()); 
            return; 
        }

        for(int i = 0; i < hashMap.get(digits.charAt(index)).size(); i++){
            sb.append(hashMap.get(digits.charAt(index)).get(i)); 
            dfs(digits, answer, index + 1, sb); 
            sb.deleteCharAt(sb.length()-1); 
        }
    }
}